### PR TITLE
apply amp bf16 on textual inversion

### PR DESCRIPTION
**Pull request by jiqing-feng**
Hi, @patrickvonplaten  @patil-suraj  . I tried to use `mixed_precision` to accelerate finetune but found that the time costs are the same between fp32 and bf16. The accelerator only works for text encoder but the unet model takes a mount of time, so I use amp bf16 to optimize unet. The performance is remarkable(around 1.55x speed-up) and the generated images are also acceptable. If you are interested with this idea, we can discuss further about this patch.

### CODE REVIEW

File: examples/textual_inversion/README.md [hunk 1]
```diff
@@ -69,6 +69,7 @@ accelerate launch textual_inversion.py \
 ```
 
 A full training run takes ~1 hour on one V100 GPU.
+Add flag `--amp_data_type=bf16` if you want to use bf16 for unet forward to accelerate finetune. A full training by this flag run takes ~30 min on one V100 GPU.
```
Comment (yao-matrix) [inline examples/textual_inversion/README.md:72]: V100 doesn't have BF16 support, BF16 support in GPU is since A100.

File: examples/textual_inversion/textual_inversion.py [hunk 2]
```diff
@@ -549,18 +556,23 @@ def main():
...
                     raise ValueError(f"Unknown prediction type {noise_scheduler.config.prediction_type}")
+                    
+                # Predict the noise residual
+                if args.amp_data_type == "bf16":
+                    with torch.autocast(device_type=str(accelerator.device), dtype=torch.bfloat16):
+                        model_pred = unet(noisy_latents, timesteps, encoder_hidden_states).sample
+                        loss = F.mse_loss(model_pred, target, reduction="none").mean([1, 2, 3]).mean()
+                else:
+                    model_pred = unet(noisy_latents, timesteps, encoder_hidden_states).sample
+                    loss = F.mse_loss(model_pred, target, reduction="none").mean([1, 2, 3]).mean()
```
Comment (patil-suraj) [inline examples/textual_inversion/textual_inversion.py:574]: Instead of using `autocrats` here, we could directly cast `unet` and `van` to low precision if `mixed_precision` is turned on as it's done in dreambooth script cf https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/train_dreambooth.py#LL566

This way we don't need to add any new argument.

**Review by patil-suraj — COMMENTED** [review]
Thanks for the PR, just left a comment.

**Comment by patrickvonplaten** [comment]
@patil-suraj 

Ideally I'd really like to have `accelerate` handle all of this. Currently I don't fully understand why we cannot have accelerate change the precision. E.g. the following works:

```python
from diffusers import AutoencoderKL
from accelerate import Accelerator
import torch

print(torch.cuda.is_available())

accelerator = Accelerator()

model = AutoencoderKL()

for param in model.parameters():
    param.requires_grad = False

accelerator.prepare(model)

print(model.device)
```

File: examples/textual_inversion/README.md (hunk 1 above)
Comment (pcuenca) [reply examples/textual_inversion/README.md:72]: I think it works on any Ampere cards? (Including 3090, A10, A5000, A6000)

File: examples/textual_inversion/textual_inversion.py (hunk 2 above)
Comment (jiqing-feng) [reply examples/textual_inversion/textual_inversion.py:574]: Thanks for your comments. We know that `.to` and `.autocast` are different ways since `to` convert all parameters to bf16 and `.autocast` may keeps come operations with fp32 (https://pytorch.org/docs/stable/amp.html). Therefore, the second way is more likely to keep the origin model's (fp32) accuracy.

File: examples/textual_inversion/textual_inversion.py [hunk 3]
```diff
@@ -191,6 +191,13 @@ def parse_args():
             "and an Nvidia Ampere GPU."
         ),
     )
+    parser.add_argument(
```
[the remark this answers is not in the document]
Comment (jiqing-feng) [reply examples/textual_inversion/textual_inversion.py:194]: This is because I can only get fp32 tensors in the output of unet, vae, text_encoder and loss even if I chose mixed_precision=bf16 in both accelerate config and command line. The output tensor of unet is bf16 when I apply autocast.   It might be my mistakes and I will check it. It would be nice if you can have a try to double check. Thanks.

File: examples/textual_inversion/textual_inversion.py (hunk 2 above)
Comment (patil-suraj) [reply examples/textual_inversion/textual_inversion.py:574]: You are right @jiqing-feng ! But here the `unet` and `vae` are not trained and are always in inference mode, so using `.to` here shouldn't hurt the accuracy. IMO `autocast` is necessary for the model being trained, which in this case is `text_encoder` which already get's wrapped in that context in `accelerate`.

**Comment by patil-suraj** [comment]
Hey @jiqing-feng !

There are two ways to go about this:

1.  As Patrick said, we can wrap all models in `accelerate.prepare` so when mixed precision is enabled, accelerate will automatically add the `autocast` wrapper. It's not necessary to add the extra `amp_data_type` argument.
2.  As done in other scripts like dreambooth and text_to_image, cast the weights of non-trainable models to half-precision. This is safe here as the unet and vae are always used in inference model so casting their weights with `.to` should not affect anything.


I tried the first option and it works and gives speed-up in mixed-precision but it takes extra 2-3GB memory as autocast creates some copies. So I would recommend the second approach, as it's already working well for dreambooth. Could you please update the PR like that, for reference see https://github.com/huggingface/diffusers/blob/ae4112d2bbfa363f2f3049daad54f0fb89c34499/examples/dreambooth/train_dreambooth.py#LL566-L577

and  https://github.com/huggingface/diffusers/blob/ae4112d2bbfa363f2f3049daad54f0fb89c34499/examples/dreambooth/train_dreambooth.py#L614-L618

**Review by patil-suraj — CHANGES_REQUESTED** [review]
Hey @jiqing-feng !

There are two ways to go about this:

1.  As Patrick said, we can wrap all models in `accelerate.prepare` so when mixed precision is enabled, accelerate will automatically add the `autocast` wrapper. It's not necessary to add the extra `amp_data_type` argument.
2.  As done in other scripts like dreambooth and text_to_image, cast the weights of non-trainable models to half-precision. This is safe here as the unet and vae are always used in inference model so casting their weights with `.to` should not affect anything.


I tried the first option and it works and gives speed-up in mixed-precision but it takes extra 2-3GB memory as autocast creates some copies. So I would recommend the second approach, as it's already working well for dreambooth. Could you please update the PR like that, for reference see https://github.com/huggingface/diffusers/blob/ae4112d2bbfa363f2f3049daad54f0fb89c34499/examples/dreambooth/train_dreambooth.py#L566-L577

and  https://github.com/huggingface/diffusers/blob/ae4112d2bbfa363f2f3049daad54f0fb89c34499/examples/dreambooth/train_dreambooth.py#L614-L618

**Comment by jiqing-feng** [comment]
@patil-suraj @patrickvonplaten Thanks for your comments! The second approach is great and I have applied it in this PR. Would you please help me to review it? Thanks!

**Comment by yao-matrix** [comment]
@kding1 this is the PR we mentioned, to upstream BF16 common practices to textual inversion, which will benefit Intel 4th generation Xeon platform also.

**Review by patrickvonplaten — APPROVED** [review]
Thanks a lot @jiqing-feng for bearing with us through the PR! This looks now very nice to me :-) 
Think we can merge it!
